packages/remix-dev/config/defaults/entry.server.spa.tsx

```diff
@@ -9,9 +9,9 @@ export default function handleRequest(
   responseHeaders: Headers,
   remixContext: EntryContext
 ) {
-  let html = renderToString(
-    <RemixServer context={remixContext} url={request.url} />
-  );
+  let html =
+    "<!DOCTYPE html>\n" +
+    renderToString(<RemixServer context={remixContext} url={request.url} />);
```

Comment:
We probably shouldn't do this blindly.  There is no guarantee that root returns a complete `<html>` document.  It's possible (in SSR and SPA) to hydrate only a portion of the document by returning a `<div>` from `root.tsx` and updating `entry.client.tsx` to hydrate a specific element.  It's not the recommended usage - but we do like to leave as much control in the app developers hands and make minimal internal assumptions.

I think it would probably be safe to do this conditionally though?

```suggestion
  let html = renderToString(
    <RemixServer context={remixContext} url={request.url} />
  );
  if (html.startsWith('<html')) {
    html = "<!DOCTYPE html>\n" + html;
  }
```

Reply:
That's a very good point